### ISSUE TITLE
Issue with Residual Prompt Injection in Vision Transformer Block in the code

### ISSUE THREAD

**Issue by Viskkk** [aimagelab/mammoth#57]
Thank you for building Mammoth, which has helped me a lot. However, I have been reviewing the implementation of the Vision Transformer (ViT) in your codebase and noticed that the Block class in vision_transformer.py has a prompts argument in its forward method. The code looks like this:
class Block(MammothViTBlock):
    def forward(self, x, prompts=None):
        x = x + self.drop_path1(self.ls1(self.attn(self.norm1(x), prompts)))
        x = x + self.drop_path2(self.ls2(self.mlp(self.norm2(x))))
        return x
However, after further inspecting the Attention class, which is defined in vit.py, I did not find any clear injection of the prompts parameter in its forward method. The Attention class’s forward function appears to handle only the typical attention mechanism and does not seem to account for semantic prompt injection.
Could you kindly clarify where or how the semantic prompts are injected into the transformer block, if at all? Is there a specific part of the code or a class that I might be missing which handles this prompt-based modification?
I would greatly appreciate your guidance on this matter.
Thank you in advance for your time and for the excellent work on this project.

**Comment by loribonna** [comment]
Hi @Viskkk , thank you for your interest! The layers in vit.py define only a basic structure of the ViT, so they follow the default behaviour of the vit and do not apply prompts.
  
When porting the code from timm, I decided to make my life easier and make all layers accept and propagate an additional `**kwargs` (as you have seen for the Attention class), so that they would have been easier to extend. 

The actual implementation of the semantic prompts for STAR-Prompt is in `models/starprompt_utils/vision_transformer.py`, specifically the class [ResidualPromptAttention](https://github.com/aimagelab/mammoth/blob/3d6fc4b0645734e5d8c416293efadc44f3032382/models/star_prompt_utils/vision_transformer.py#L9) applies the residual prompts and the class [ResidualPromptAttention](https://github.com/aimagelab/mammoth/blob/3d6fc4b0645734e5d8c416293efadc44f3032382/models/star_prompt_utils/vision_transformer.py#L53) redefines the ViT so that it generates the prompts for each block (in the `forward_features` method).

**Comment by Viskkk** [comment]
Thank you very much for your reply! You have resolved my issue perfectly.

**Comment by Viskkk** [comment]
Could I ask one more question? In the first stage of training with STAR-prompt, did you freeze the prompts learned from earlier tasks simply by using:
with torch.no_grad(): 
    past_keys = self.compute_keys(0, frozen_past_classes)
Or was there another method used for freezing the earlier prompts?

**Comment by Viskkk** [comment]
Could I ask one more question? In the first stage of training with STAR-prompt, did you freeze the prompts learned from earlier tasks simply by using the code below when computing the keys?
##
with torch.no_grad():
    past_keys = self.compute_keys(0, frozen_past_classes)
##
Or was there another method used for freezing the earlier prompts?

**Comment by loribonna** [comment]
No problem. We used both that code in line 191 of `models/star_prompt_utils/first_stage_starprompt.py` and we detached the retrieved keys in line 194 with `keys = torch.cat((past_keys.detach(), cur_keys), dim=0)`.

I don't think both lines are necessary to freeze the parameters, either one would have probably been enough, but we wanted to be sure.

If you want to run the first stage and learn all the keys during all the tasks, I suggest you just compute the keys as:
`keys = self.compute_keys(0, cur_classes)`

**Comment by Viskkk** [comment]
Thank you very much!

**Comment by Viskkk** [comment]
However, when I run the StarPrompt code, the training process often randomly freezes at a certain point and stops progressing, which is something I usually don't encounter when running other code.Have you encountered this issue before, and how can I resolve it?

**Comment by loribonna** [comment]
Does it stop randomly in the middle of a epoch or at the start/end of a task? If it is at the start of a task it is probably something about the loading of the data (the initialization of the dataloaders). If it is in the middle of a task is might be more tricky. 

Can you share the output and the string used to launch?

**Comment by Viskkk** [comment]
The issue occurs either randomly at the end of an epoch or right at the start of the evaluation phase, just like what’s shown in the image. As for launching the script, I directly added the following to the main.py file:
sys.argv = [
    'main.py',
    '--model', 'first_stage_starprompt',
    '--lr', '0.002',
    '--n_epochs', '20',
    '--gr_mog_n_iters_first_stage', '200',
    '--lambda_ortho_first_stage', '30',
    '--dataset', 'seq-imagenet-r',
]
Afterward, I simply run the command python main.py in the terminal.
![1](https://github.com/user-attachments/assets/44af92a8-7abc-4f65-a4b7-b74be0a80b63)

![2](https://github.com/user-attachments/assets/bd237fab-81ab-4d32-b6df-185c28df53a1)

**Comment by loribonna** [comment]
I'm not sure, I tried on my machine and it seems to run. 
I think it may be an issue with the workers since it occurs at the beginning of an epoch or evaluation. It seems from your screen that the number of workers is set to 0, which is strange since there is no `--num_workers=0` in your parameters. Can you try running with `--num_workers=2`?

**Comment by Viskkk** [comment]
I tried to set --num_workers=2 and it worked.
Thank you very much for your patient response!

**Comment by loribonna** [comment]
I think these are the dataloaders created during the generative replay process [in line 96 of models/star_prompt_utils/first_stage_model.py](https://github.com/aimagelab/mammoth/blob/3d6fc4b0645734e5d8c416293efadc44f3032382/models/star_prompt_utils/first_stage_model.py#L96). 

We forced num_workers to 0 there to make training faster (the few examples to load are already in local memory so most of the time will be spent creating the worker processes with num_workers>0). You might try changing the line to:
`return create_seeded_dataloader(self.args, TensorDataset(features, labels), num_workers=2, batch_size=self.args.batch_size_gr, shuffle=True)`

**Comment by Viskkk** [comment]
I made the modifications according to your suggestions, but the following error occurred. Could you please advise on how I should proceed with further adjustments?

![image](https://github.com/user-attachments/assets/ead2f098-2db9-4b3b-b0b0-6361effe5558)

**Comment by loribonna** [comment]
I confirm that setting multiple workers at that point will make the GPU crash... didn't know that.

Maybe you can try limiting the number of threads spawned by the dataloaders with `torch.set_num_threads(0)` (after reverting back the change I suggested earlier). 

Sorry I can't be more precise but it seems to be an issue with how Mammoth handles machines with many cpus and I don't have a way to reproduce this locally.

**Comment by Viskkk** [comment]
Finally,I found it work when I set --num_workers=0. Thanks

**Comment by loribonna** [comment]
Great! Just out of curiosity, what kind of GPU do you have? 
I ask just because on my 1080 i average ~25 ep/h while you seem to be >1000!

**Comment by Viskkk** [comment]
hhhh,I used an 4090.

**Comment by Viskkk** [comment]
Hhhh,I use single-card training on 4090

**Comment by loribonna** [comment]
oh wow we really need to get our hands on some of them. Thank you, I will close the issue.
